File: app/src/main/java/com/merqury/aspu/services/news/NewsService.java
package com.merqury.aspu.services.news;

import android.annotation.SuppressLint;

import androidx.annotation.OptIn;
import androidx.compose.foundation.ExperimentalFoundationApi;

import com.merqury.aspu.enums.NewsCategoryEnum;
import com.merqury.aspu.services.appconfig.AppConfig;
import com.merqury.aspu.ui.navfragments.news.NewsStates;

import java.util.List;

public class NewsService {
    private static final String faculty_header = "faculties/";
    private static final String urlForEverything = "agpu.net/struktura-vuza/%s/news/news.php?PAGEN_1=%d";
    private final static String urlAgpuNews = "agpu.net/news.php";
    private final static List<String> nonStandardCategories;

    static {
        nonStandardCategories = List.of(
                "educationaltechnopark",
                "PedagogicalQuantorium"
        );
    }

    @SuppressLint("DefaultLocale")
    @OptIn(markerClass = ExperimentalFoundationApi.class)
    public static String urlForCurrentFaculty() {
        String faculty = NewsStates.INSTANCE.getSelectedFaculty().name();
        int page = NewsStates.INSTANCE.getPagerState().getCurrentPage();
        if(NewsStates.INSTANCE.getSelectedFaculty() == NewsCategoryEnum.agpu)
            return urlAgpuNews+"?PAGEN_1="+(page+1);
        return String.format(urlForEverything, nonStandardCategories.contains(NewsStates.INSTANCE.getSelectedFaculty().name()) ? faculty : faculty_header + faculty, (page+1));
    }
}

File: app/src/main/java/com/merqury/aspu/services/news/NewsService.kt
package com.merqury.aspu.services.news

import com.merqury.aspu.enums.NewsCategoryEnum
import com.merqury.aspu.services.api.news.GetNewsService
import com.merqury.aspu.services.api.news.models.FullArticle
import com.merqury.aspu.services.api.news.models.NewsResponse
import com.merqury.aspu.ui.async
import com.merqury.aspu.ui.printlog
import java.net.UnknownHostException

fun getNews(
    pageNumber: Int,
    selectedFaculty: NewsCategoryEnum,
    onError: (String) -> Unit,
    onSuccess: (NewsResponse) -> Unit
) {
    async {
        try {
            val newsService = GetNewsService.getInstance()
            val response = newsService.getArticlesByFaculty(selectedFaculty.name, pageNumber)
            onSuccess(response)
        } catch (e: Exception) {
            if (e is UnknownHostException) {
                onError("Нет подключения к сети Интернет")
            } else {
                printlog("NEWS FETCH ERROR: $e")
                onError("Неизвестная ошибка")
            }
        }
    }
}

fun getNewsArticle(
    faculty: NewsCategoryEnum,
    id: Int,
    onSuccess: (FullArticle) -> Unit,
    onError: (String) -> Unit
) {
    async {
        try {
            val newsService = GetNewsService.getInstance()
            val response = newsService.getArticleById(faculty.name, id)
            onSuccess(response)
        } catch (e: Exception) {
            if (e is UnknownHostException) {
                onError("Нет подключения к сети Интернет")
            } else {
                printlog("NEWS ARTICLE FETCH ERROR: $e")
                onError("Неизвестная ошибка")
            }
        }
    }
}


//@OptIn(ExperimentalFoundationApi::class)
//fun urlForCurrentFaculty(): String {
//    return when (selectedFaculty.value.name) {
//        "agpu" -> "agpu.net/news.php?PAGEN_1=${pagerState.value.currentPage}"
//        "educationaltechnopark" -> "www.agpu.net/struktura-vuza/educationaltechnopark/news/news.php?PAGEN_1=${pagerState.value.currentPage}"
//        "PedagogicalQuantorium" -> "www.agpu.net/struktura-vuza/PedagogicalQuantorium/news/news.php?PAGEN_1=${pagerState.value.currentPage}"
//        else -> "agpu.net/struktura-vuza/faculties/${selectedFaculty.value.name}/news/news.php?PAGEN_1=${pagerState.value.currentPage}"
//    }
//}